### PR TITLE
11-Refactor-the-simulator-to-be-multi-threaded pull request

### PR DESCRIPTION
**Pull request by Nadim-Muhammad**
Created a multi-threaded underlying price simulator and fixed a CRUCIAL BUG in simulator.cpp (and theorethical values..) which I did not saw earlier, which gave us wrong outputs. Im 99% confidend it should be correct now. @Staneesh Please check and merge if it's good.

### CODE REVIEW

File: src/simulator.cpp [hunk 1]
```diff
@@ -19,14 +19,55 @@ void StockSimulator::runSimulation() {
...
+    int iterationsPerThread = static_cast<int>(iter / numThreads);
+    for (unsigned int i = 0; i < numThreads; ++i) {
+        int start = i * iterationsPerThread;
+        int end = (i == numThreads - 1) ? iter : (i + 1) * iterationsPerThread;
+        threads.emplace_back(simulationThread, start, end);
+    }
+
+    // Wait for all threads to finish
+    for (auto& thread : threads) {
+        thread.join();
```
Comment (Staneesh) [inline src/simulator.cpp:65]: Does `thread.join()` return some value that could be checked for determining if `join()` succeeded?

File: src/simulator.h [hunk 2]
```diff
@@ -21,6 +22,8 @@ class StockSimulator {
 
     void runSimulation();
 
+    void runSimulationParallel();
```
Comment (Staneesh) [inline src/simulator.h:25]: Please have a prompt in `main` asking the user if he wants to run tests or the single-threaded simulation or the multi-threaded simulation. Please also add a prompt for the maximum number of threads that the user wants to use. In the limit, the time spent running single-threaded simulation should equal 1-thread variant of the multi-threaded simulation. 

Limiting number of threads used is super useful if you don't want to overload your machine and keep doing other tasks.

**Review by Staneesh — CHANGES_REQUESTED** [review]
Hey, left some notes below 🌴

**Comment by Nadim-Muhammad** [comment]
Updated the 11-Refactor-the-simulator-to-be-multi-threaded branch, commented the code and did things you mentioned. Merged runSimulation() and runSimulationParallel() into one and added propmt to choose how many cores does user want to use. Check code again and merge if it's good.

**Pull request merged**